README.md
# split-diff [![Installs!](https://img.shields.io/apm/dm/split-diff.svg?style=flat-square)](https://atom.io/packages/split-diff) [![Version!](https://img.shields.io/apm/v/split-diff.svg?style=flat-square)](https://atom.io/packages/split-diff) [![License](https://img.shields.io/apm/l/split-diff.svg?style=flat-square)](https://github.com/mupchrch/split-diff/blob/master/LICENSE.md)

Diffs text between two split panes. New panes are created if less than two panes exist upon run of the package.

\*\* **Supports diffing recent git changes!** \*\*

![split-diff in action](https://github.com/mupchrch/split-diff/raw/master/demo.gif)

## Minimap Plugin

Get the [split-diff minimap plugin](https://atom.io/packages/minimap-split-diff) to make it easier to spot differences.

## Usage

### Commands

You can call any of these commands from your own custom keybinding, simply override the command in your keymap.cson!

| Command | Effect | Default Keymaps |
| ------- | ------ | --------------- |
| Toggle `split-diff:toggle` | Toggles a diff of the text between two side-by-side panes. Creates panes as necessary. Displays git changes if there is a repository found. | `ctrl-alt-t`,<br>`ctrl-alt-d` (_linux_) |
| Enable `split-diff:enable` | Enables a diff of the text between two side-by-side panes. Creates panes as necessary. Displays git changes if there is a repository found. | ... |
| Disable `split-diff:disable` | Disables a diff. Removes any panes that were created by this package. | ... |
| Set Ignore Whitespace `split-diff:set-ignore-whitespace` | Toggles the Ignore Whitespace setting. | ... |
| Set Auto Diff `split-diff:set-auto-diff` | Toggles the Auto Diff setting. | ... |
| Move to Next Diff `split-diff:next-diff` | Highlights the next difference. Moves the cursor so it can be easily copied. | `ctrl-alt-n` |
| Move to Previous Diff `split-diff:prev-diff` | Highlights the previous difference. Moves the cursor so it can be easily copied. | `ctrl-alt-p` |
| Copy to Right `split-diff:copy-to-right` | Copies all differences containing a cursor to the right pane. | `ctrl-alt-.` |
| Copy to Left `split-diff:copy-to-left` | Copies all differences containing a cursor to the left pane. | `ctrl-alt-,` |

### Settings

* **Auto Diff** - Automatically recalculates the diff when one of the editors changes.
* **Show Word Diff** - Diffs the words between each line when this box is checked.
* **Ignore Whitespace** - Will not diff whitespace when this box is checked.
* **Mute Notifications** - Mutes all warning notifications when this box is checked.
* **Remove Soft Wrap** - Removes soft wrap during diff - restores when finished.
* **Hide Docks** - Hides all docks (Tree View, Github, etc) during diff - shows when finished.
* **Sync Scrolling** - Syncs the scrolling of the editors.
#### Colors
* **Added Color Side** - Which editor (left or right) to highlight as added. The opposite editor will be highlighted as removed.
* **Override Highlight Colors** - Whether to override diff colors derived from the current syntax theme.
* **Added Custom Color** - The color that will be used when overriding added highlight colors.
* **Removed Custom Color** - The color that will be used when overriding removed highlight colors.

### Service API
Packages can consume the split-diff service to do things like enable a diff between two editors.

```js
/**
 * Getter for the marker layers of each editor being diffed.
 * @return {Promise} A promise that resolves to an object containing the marker layers.
 */
getMarkerLayers();

/**
 * Enables split-diff between the two given editors.
 * @param {TextEditor} editor1 - The left editor.
 * @param {TextEditor} editor2 - The right editor.
 * @param {object} options - Options to override any package setting.
 */
diffEditors(editor1, editor2, options);

/**
 * Disables split-diff.
 */
disable();
```


CHANGELOG.md
## 1.6.1 - 2019-09-27
* Fixed "Diff with Active File" opens blank document #165

## 1.6.0 - 2019-01-24
* Added ability to disable automatic diff #139
* Changed loading indicator to appear in footer UI instead of as a modal that covered both editors #149
* Fixed toggling "Ignore Whitespace" from Atom settings didn't update the checkbox in the footer UI
* Fixed "Diff with Active File" not highlighting syntax in newly created tab (missed part of #151)
* Fixed "Diff with Active File" of new, unsaved tab creating a new tab with editor hooked up to shared text buffer #160
* Fixed "Remove Soft Wrap" setting not working for quick git diff #143

## 1.5.4 - 2019-01-08
* Fixed bug with setting the grammar for second editor #151 - thanks hSaria!

## 1.5.3 - 2018-10-10
* Fixed syntax highlighting not working for automatically opened editors #141

## 1.5.2 - 2017-12-29
* Fixed infinite scroll loop when performing a next/prev diff command #136

## 1.5.1 - 2017-08-29
* Fixed selected line losing its background color with some themes - thanks skylerlee!

## 1.5.0 - 2017-08-17
* Added setting to automatically turn off soft wrap during diff and restore it on disable #101
* Added git submodule support #112
* Changed octocat loading icon to be a more generic (full CSS) loading spinner #126
* Fixed error when choosing "Diff with Active File" on an editor tab created by this package #125
* Fixed next/prev selection number showing "1 / 0" when clicking next with zero differences

## 1.4.4 - 2017-08-16
* Changed "Hide Tree View" setting to "Hide Docks" which fixes tree view re-showing while typing issue #125

## 1.4.3 - 2017-06-22
* Fixed issue scrolling and splitting panes in Atom v1.19.0-beta1 #120
* Fixed deprecation warning when turning on the package #102

## 1.4.2 - 2017-06-21
* Changed toggle keybinding back to ctrl-alt-t for windows and mac #117 - thanks ooJerryLeeoo!

## 1.4.1 - 2017-06-11
* Fixed ctrl-alt-t keybinding conflict in Linux with terminal shortcut (changed to ctrl-alt-d) #98
* Fixed uncaught type error for "Diff With Active File" when using Nuclide package #85
* Fixed error for "Diff With Active File" when selected for a tab not saved on disk

## 1.4.0 - 2017-05-17
* Added option to override diff highlight colors #78
* Added service hook to diff two editors with the ability to override a user's settings without overwriting them (to be used from other packages) #69
* Added service hook to disable the package
* Fixed "Hide Tree View" setting blindly opening tree-view regardless of if it was already hidden

## 1.3.1 - 2017-04-21
* Fixed quick diff adding text editor in panes outside of center work area #104

## 1.3.0 - 2017-03-26
* Added highlight color on click of a diff chunk to indicate selection
* Changed warning message when attempting to copy left/right with no selections
* Fixed uncaught type error #94 - thanks littlebee!
* Fixed being unable to copy an empty chunk to other editor #95

## 1.2.1 - 2017-03-17
* Fixed uncaught type error #93 - thanks Greenek!
* Fixed uncaught reference error #96 - thanks Greenek!

## 1.2.0 - 2017-02-20
* Added option to automatically hide Tree View while diffing #82

## 1.1.4 - 2017-02-12
* Rerelease changes made for v1.1.2 but scoped to Atom v1.14 and greater

## 1.1.3 - 2017-02-12
* Fixed error in Atom v1.13 due to MarkerLayer bug (rollback changes) #92

## 1.1.2 - 2017-02-09
* Added service API for getting marker layers and related metadata
* Fixed marker deprecation warnings due to an optional property removal in 1.8 #87

## 1.1.1 - 2017-01-09
* Fixed styling deprecation warnings due to removal of shadow DOM #86

## 1.1.0 - 2016-11-29
* Added ability to diff active file with right click in tree view or tab header #9
* Fixed toggling ignore whitespace via command not updating its state in the footer bar
* Fixed issue where one line editor would always pop notification about differing line endings
* Fixed word diff not working for line diffed against an empty line
* Fixed wrong current selection counter after copying last chunk to other editor
* Fixed poor user experience when auto opening text editors in new panes (Now tries to reuse panes more often instead of greedily creating new ones)
* Large code refactor

## 1.0.6 - 2016-10-24
* Fixed next/prev diff highlight skipping a selection after copying to right/left
* Added keybinding for copy to left/right and scoped it to just the editors #73

## 1.0.5 - 2016-10-05
* Added option to mute warning notifications
* Updated option for scroll syncing to be able to turn off all syncing #68
* Changed package to no longer disable soft wrap while diffing #71

## 1.0.4 - 2016-08-13
* Added keybinding for toggle #64 - thanks moritamori!
* Changed next/prev diff command to toggle package if not enabled #66
* Fixed diff hanging when diffing files containing json/xml #65
* Fixed uncaught exception when pressing copy chunk button before diff finished
* Fixed uncaught exception when pressing next/prev button before diff finished

## 1.0.3 - 2016-06-20
* Reintroduced one second delay to the loading modal #62

## 1.0.2 - 2016-06-03
* Actually fixed line ending normalization for git repo #39 - thanks IceOnFire!

## 1.0.1 - 2016-06-01
* Fixed Uncaught TypeError when trying to highlight word diff #60
* Fixed new pane not using same line endings as first pane for git repo #39
* Fixed last chunk not being the same size which caused misalignment of editors #58

## 1.0.0 - 2016-05-26
* Added a user interface!
* Added help notification when user fails the copy command
* Added new pane opened by package will use same line endings as first pane (except for git panes) #39
* Removed notifications on enable/disable #57
* Changed next/prev diff command to move cursor with it so copy commands are easy #51
* Fixed error when trying to copy last chunk when it didn't have a match #50
* Fixed ignore whitespace not affecting the diff outcome #52
* Fixed Uncaught TypeError when trying to highlight word diff #49

lib/build-lines.js
'use babel'

module.exports = class DiffViewEditor {
  _editor: Object;
  _markers: Array<atom$Marker>;
  _currentSelection: Array<atom$Marker>;

  constructor(editor) {
    this._editor = editor;
    this._markers = [];
    this._currentSelection = null;
    this._oldPlaceholderText = editor.getPlaceholderText();
    editor.setPlaceholderText('Paste what you want to diff here!');
    // add split-diff css selector to editors for keybindings #73
    atom.views.getView(this._editor).classList.add('split-diff');
  }

  /**
   * Creates a decoration for an offset. Adds the marker to this._markers.
   *
   * @param lineNumber The line number to add the block decoration to.
   * @param numberOfLines The number of lines that the block decoration's height will be.
   * @param blockPosition Specifies whether to put the decoration before the line or after.
   */
  _addOffsetDecoration(lineNumber, numberOfLines, blockPosition): void {
    var element = document.createElement('div');
    element.className += 'split-diff-offset';
    // if no text, set height for blank lines
    element.style.minHeight = (numberOfLines * this._editor.getLineHeightInPixels()) + 'px';

    var marker = this._editor.markScreenPosition([lineNumber, 0], {invalidate: 'never', persistent: false});
    this._editor.decorateMarker(marker, {type: 'block', position: blockPosition, item: element});
    this._markers.push(marker);
  }

  /**
   * Adds offsets (blank lines) into the editor.
   *
   * @param lineOffsets An array of offsets (blank lines) to insert into this editor.
   */
  setLineOffsets(lineOffsets: any): void {
    var offsetLineNumbers = Object.keys(lineOffsets).map(lineNumber => parseInt(lineNumber, 10)).sort((x, y) => x - y);

    for (var offsetLineNumber of offsetLineNumbers) {
      if (offsetLineNumber == 0) {
        // add block decoration before if adding to line 0
        this._addOffsetDecoration(offsetLineNumber-1, lineOffsets[offsetLineNumber], 'before');
      } else {
        // add block decoration after if adding to lines > 0
        this._addOffsetDecoration(offsetLineNumber-1, lineOffsets[offsetLineNumber], 'after');
      }
    }
  }

  /**
   * Creates markers for line highlights. Adds them to this._markers. Should be
   * called before setLineOffsets since this initializes this._markers.
   *
   * @param changedLines An array of buffer line numbers that should be highlighted.
   * @param type The type of highlight to be applied to the line.
   */
  setLineHighlights(changedLines: Array<number> = [], highlightType: string): void {
    var highlightClass = 'split-diff-' + highlightType;
    for (var i=0; i<changedLines.length; i++) {
      this._markers.push(this._createLineMarker(changedLines[i][0], changedLines[i][1], highlightClass));
    }
  }

  /**
   * Creates a marker and decorates its line and line number.
   *
   * @param startLineNumber A buffer line number to start highlighting at.
   * @param endLineNumber A buffer line number to end highlighting at.
   * @param highlightClass The type of highlight to be applied to the line.
   *    Could be a value of: ['split-diff-insert', 'split-diff-delete',
   *    'split-diff-select'].
   * @return The created line marker.
   */
  _createLineMarker(startLineNumber: number, endLineNumber: number, highlightClass: string): atom$Marker {
    var marker = this._editor.markBufferRange([[startLineNumber, 0], [endLineNumber, 0]], {invalidate: 'never', persistent: false, class: highlightClass})

    this._editor.decorateMarker(marker, {type: 'line-number', class: highlightClass});
    this._editor.decorateMarker(marker, {type: 'line', class: highlightClass});

    return marker;
  }

  /**
   * Highlights words in a given line.
   *
   * @param lineNumber The line number to highlight words on.
   * @param wordDiff An array of objects which look like...
   *    added: boolean (not used)
   *    count: number (not used)
   *    removed: boolean (not used)
   *    value: string
   *    changed: boolean
   * @param type The type of highlight to be applied to the words.
   */
  setWordHighlights(lineNumber: number, wordDiff: Array<any> = [], type: string, isWhitespaceIgnored: boolean): void {
    var klass = 'split-diff-word-' + type;
    var count = 0;

    for (var i=0; i<wordDiff.length; i++) {
      if (wordDiff[i].value) { // fix for #49
        // if there was a change
        // AND one of these is true:
        // if the string is not spaces, highlight
        // OR
        // if the string is spaces and whitespace not ignored, highlight
        if (wordDiff[i].changed
          && (/\S/.test(wordDiff[i].value)
          || (!/\S/.test(wordDiff[i].value) && !isWhitespaceIgnored))) {
          var marker = this._editor.markBufferRange([[lineNumber, count], [lineNumber, (count + wordDiff[i].value.length)]], {invalidate: 'never', persistent: false, class: klass})

          this._editor.decorateMarker(marker, {type: 'highlight', class: klass});
          this._markers.push(marker);
        }
        count += wordDiff[i].value.length;
      }
    }
  }

  /**
   * Destroys all markers added to this editor by split-diff.
   */
  destroyMarkers(): void {
    for (var i=0; i<this._markers.length; i++) {
      this._markers[i].destroy();
    }
    this._markers = [];

    this.deselectAllLines();
  }

  /**
   * Destroys the instance of the DiffViewEditor and cleans up after itself.
   */
  destroy(): void {
      this.destroyMarkers();
      this._editor.setPlaceholderText(this._oldPlaceholderText);
      // remove split-diff css selector from editors for keybindings #73
      atom.views.getView(this._editor).classList.remove('split-diff')
  }

  /**
   * Not added to this._markers because we want it to persist between updates.
   *
   * @param startLine The line number that the selection starts at.
   * @param endLine The line number that the selection ends at (non-inclusive).
   */
  selectLines(startLine: number, endLine: number): void {
    // don't want to highlight if they are the same (same numbers means chunk is
    // just pointing to a location to copy-to-right/copy-to-left)
    if (startLine < endLine) {
      this._currentSelection = this._createLineMarker(startLine, endLine, 'split-diff-selected');
    }
  }

  /**
   * Destroy the selection markers.
   */
  deselectAllLines(): void {
    if (this._currentSelection) {
      this._currentSelection.destroy();
      this._currentSelection = null;
    }
  }

  /**
   * Used to test whether there is currently an active selection highlight in
   * the editor.
   *
   * @return A boolean signifying whether there is an active selection highlight.
   */
  hasSelection(): boolean {
    if(this._currentSelection) {
        return true;
    }
    return false;
  }

  /**
   * Enable soft wrap for this editor.
   */
  enableSoftWrap(): void {
    try {
      this._editor.setSoftWrapped(true);
    } catch (e) {
      //console.log('Soft wrap was enabled on a text editor that does not exist.');
    }
  }

  /**
   * Removes the text editor without prompting a save.
   */
  cleanUp(): void {
    // if the pane that this editor was in is now empty, we will destroy it
    var editorPane = atom.workspace.paneForItem(this._editor);
    if (typeof editorPane !== 'undefined' && editorPane != null && editorPane.getItems().length == 1) {
      editorPane.destroy();
    } else {
      this._editor.setText('');
      this._editor.destroy();
    }
  }

  /**
   * Finds cursor-touched line ranges that are marked as different in an editor
   * view.
   *
   * @return The line ranges of diffs that are touched by a cursor.
   */
  getCursorDiffLines(): any {
    var cursorPositions = this._editor.getCursorBufferPositions();
    var touchedLines = [];

    for (var i=0; i<cursorPositions.length; i++) {
      for (var j=0; j<this._markers.length; j++) {
        var markerRange = this._markers[j].getBufferRange();

        if (cursorPositions[i].row >= markerRange.start.row
          && cursorPositions[i].row < markerRange.end.row) {
            touchedLines.push(markerRange);
            break;
        }
      }
    }

    // put the chunks in order so the copy function doesn't mess up
    touchedLines.sort(function(lineA, lineB) {
      return lineA.start.row - lineB.start.row;
    });

    return touchedLines;
  }

  /**
   * Used to get the Text Editor object for this view. Helpful for calling basic
   * Atom Text Editor functions.
   *
   * @return The Text Editor object for this view.
   */
  getEditor(): TextEditor {
    return this._editor;
  }
};
